docs/development/module/debugging.md

```diff
@@ -0,0 +1,146 @@
+---
+title: Debugging
+---
+
+Debugging is a necessary in all walks of software development, and blockchain is no exception. Most of the same tools used for general purpose Rust debugging also apply to Substrate. However, there are some restrictions when operating inside of a no_std environment like the Substrate runtime.
+
+
+## Substrate's Own `print` Function
+
+To facilitate the debugging of the runtime, Substrate provides extra tools for `Print` debugging (or tracing). The [`print` function](https://substrate.dev/rustdocs/master/sp_runtime/fn.print.html) can be used as a way to log the status of the runtime execution.
+
+``` rust
+use sp_runtime::print;
+```
+``` rust
+// --snip--
+pub fn do_something(origin) -> DispatchResult {
+	print("Execute do_something");
+
+	let who = ensure_signed(origin)?;
+	let my_val: u32 = 777;
+
+	Something::put(my_val);
+
+	print("After storing my_val");
+
+	Self::deposit_event(RawEvent::SomethingStored(my_val, who));
+	Ok(())
+}
+// --snip--
+```
+
+Start the chain using the `RUST_LOG` environment variable to see the print logs.
+``` sh
+RUST_LOG=runtime=debug ./target/release/node-template --dev
+```
+
+The values are printed in the terminal or the standard output every time that the
+runtime function gets called.
+``` sh
+2020-01-01 00:00:00 tokio-blocking-driver DEBUG runtime  Execute do_something
+2020-01-01 00:00:00 tokio-blocking-driver DEBUG runtime  After storing my_val
+```
+
+## Printable Trait
+
+The [`print` function](https://substrate.dev/rustdocs/master/sp_runtime/fn.print.html) works with any type that implements the [`Printable` trait](https://substrate.dev/rustdocs/master/sp_runtime/traits/trait.Printable.html). Substrate implements this trait for some types (`u8`, `u32`, `u64`, `usize`, `&[u8]`, `&str`) by default. You can also implement it for your own custom types. Here is an example of implementing it for a pallet's `Error` type.
+
+``` rust
+// The pallet's errors
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// Value was None
+		NoneValue,
+		/// Value reached maximum and cannot be incremented further
+		StorageOverflow,
+	}
+}
+
+impl<T: Trait> Printable for Error<T> {
+	fn print(&self) {
+		match self {
+			Error::NoneValue => "Unexpected value".print(),
+			Error::StorageOverflow => "Value exceeded".print(),
+			_ => "Invalid case".print(),
+		}
+	}
+}
+
+```
+
+``` rust
+impl traits::Printable for DispatchError {
+	fn print(&self) {
+		"DispatchError".print();
+		match self {
+			Self::Other(err) => err.print(),
+			Self::CannotLookup => "Can not lookup".print(),
+			Self::BadOrigin => "Bad origin".print(),
+			Self::Module { index, error, message } => {
+				index.print();
+				error.print();
+				if let Some(msg) = message {
+					msg.print();
+				}
+			}
+		}
+	}
+}
```

Comment:
This is the code from Substrate codebase where the conversion happens. Somehow it prefers to use the `error.print()` instead of the previous code where we implmented `impl<T: Trait> Printable for Error<T> `